plugins/modules/win_domain_computer.ps1

```diff
@@ -146,6 +157,73 @@ Function Add-ConstructedState($desired_state) {
   $result.changed = $true
 }
 
+Function Invoke-OfflineDomainJoin {
+  [CmdletBinding(SupportsShouldProcess=$true)]
+  param(
+    [Parameter(Mandatory=$true)]
+    [System.Collections.IDictionary]
+    $desired_state ,
+
+    [Parameter()]
+    [bool]
+    $ReturnBlob = $false ,
+
+    [Parameter()]
+    [System.IO.FileInfo]
+    $BlobPath
+  )
+
+  End {
+    $dns_domain = $desired_state.dns_hostname -replace '^[^.]+\.'
+
+    $delete = -not $BlobPath
+    if ($delete) {
+      $BlobPath = [System.IO.Path]::GetTempFileName()
+    }
+
+    $arguments = @(
+      '/PROVISION'
+      '/REUSE'  # we're pre-creating the machine normally to set other fields, then overwriting it with this
+      '/DOMAIN'
+      $dns_domain
+      '/MACHINE'
+      $desired_state.sam_account_name.TrimEnd('$')  # this machine name is the short name
+      '/MACHINEOU'
+      $desired_state.ou
+      '/SAVEFILE'
+      $BlobPath.FullName
+    )
+
+    $argstring = Argv-ToString -arguments $arguments
+    $exe = 'djoin.exe'
+    $invocation = "$exe $argstring"
+    $result.djoin = @{
+      invocation = $invocation
+    }
+    $result.odj_blob = [String]::Empty
+
+    if ($PSCmdlet.ShouldProcess($exe, $argstring)) {
+      $djoin_result = Run-Command -command $invocation
+      $result.djoin.rc = $djoin_result.rc
+      $result.djoin.stdout = $djoin_result.stdout
+      $result.djoin.stderr = $djoin_result.stderr
+
+      if ($djoin_result.rc) {
+        Fail-Json -obj $result -message "Problem running djoin.exe. See returned values."
+      }
+
+      if ($ReturnBlob) {
+        $data = Get-Content -LiteralPath $BlobPath.FullName -Encoding Unicode -Raw
```

Comment:
What's the actual format of the contents? Just wondering whether we should return the raw bytes as a base64 string instead of trying to return the string contents and removing bytes to achieve that.

Reply:
I was actually talking about this with a co-worker today. The contents are themselves a base64 encoded string, except that the file ends with a null character (which seems to cause issues trying to send that through Ansible raw). 

I could re-encode the entire file with base64 and return it, but then in order to use it, on the controller, it would need to be decoded and the null character may cause issues again.

So in considering this, I think it ends up being easier to use it this way (I added an example in the docs), but happy to consider alternatives.

Reply:
Sigh what a crappy UX for djoin but it's not like we can fix that. Personally I think we should still return the base64 encoded bytes of the file and not the massaged contents. Here are the disadvantages I see with the current process:

* The actual file is UTF-16 encoded, the user of this function would have to know to use Unicode encoding of the value when creating the file on the host to join
* The `\u0000` value on the end cannot be transferred across which means it's more manual work by the end user to rebuild the file
* This is scraping the barrel but it is also less susceptible to changes made to the format in the future. Personally I doubt MS would change it but if they do add more binary data it would be a problem here

Whereas returning the base64 string has the advantages

* Is a pure representation of the file bytes, user's know need to care about the encoding unless they want to convert to a string themselves
* Single step in PowerShell to create a file from it
* Not reliant on the format staying the same

While some of those points are sort off solved by the examples you've it's still extra knowledge they need to be aware of. If it was just that the file needs to be UTF-16 encoded then I would be fine with returning the data as is, but because it needs both the null char and UTF-16 encoding I would err on the side of base64'ing the file itself.

In this case the example in the docs would be

```yaml
- name: Join a workgroup computer to the domain
  vars:
    target_blob_file: 'C:\ODJ\blob.txt'
  ansible.windows.win_shell: |
    $blob = [Convert]::FromBase64String('{{ computer_status.obj_blob }}')
    [IO.File]::WriteAllBytes('{{ target_blob_file }}', $blob)
    & djoin.exe --% /RequestODJ /LoadFile '{{ target_blob_file }}' /LocalOS /WindowsPath "%SystemRoot%"
```

The only missing part of all this was if copy/win_copy supported setting the content has base64 data and it did all that for you. Unfortunately that won't work today and you would have to create the file on the controller manually defeating and then using win_copy making things more complicated.

Reply:
Ok yeah, that solves some of what I saw as a disadvantage of re-base64 encoding, is that even preserving the utf-16 representation, the consumer would still have to know upon doing a decode to also ensure they use utf-16. But side-stepping putting it in a string by just using the raw bytes of the decode makes that more appealing; I'll go with that.

I would consider adding the support to `win_domain_membership` but in my actual use case I don't use that module so it's tough for me to find time, especially to do the testing needed for the implementation.

I actually use this blob to perform a join on deploy of an AMI in AWS; putting a similar PowerShell snippet as above into user data (we never have to retrieve the generated admin password or do any further config before the instance joins itself to the domain and we can connect with domain creds).

I could at least write up a feature request that could serve as a guide in case someone else is interested in taking it up.

Reply:
> I would consider adding the support to win_domain_membership but in my actual use case I don't use that module so it's tough for me to find time, especially to do the testing needed for the implementation.

Totally understandable, just contributing this stuff is excellent and adds value to the module. Please don't feel pressured to add further features to `win_domain_membership`.

> I could at least write up a feature request that could serve as a guide in case someone else is interested in taking it up.

You can, my only concern is that this will just become another feature request that won't be worked on because we have similar time constraints. Definitely won't stop you from doing it though, just don't feel obliged to create it.

Reply:
This all worked fine; the content is of course a lot larger now but otherwise looks good.